presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorPlanOptimizerProvider.java

```diff
@@ -14,10 +14,67 @@
 package com.facebook.presto.spi.connector;
 
 import com.facebook.presto.spi.ConnectorPlanOptimizer;
+import com.facebook.presto.spi.function.FunctionHandle;
 
+import java.util.List;
+import java.util.Optional;
 import java.util.Set;
+import java.util.function.BiFunction;
 
-public interface ConnectorPlanOptimizerProvider
+import static java.util.Collections.emptySet;
+
+public interface ConnectorPlanOptimizerProvider<T>
 {
-    Set<ConnectorPlanOptimizer> getConnectorPlanOptimizers();
+    /**
+     * @return a list of function mappings for registration.
+     * The function translation will be used by {@link #getConnectorPlanOptimizers(TranslationContext)}
+     */
+    default Set<Class<?>> getFunctionTranslation()
+    {
+        return emptySet();
+    }
+
+    /**
+     * @param functionTranslator that maps CallExpression to TargetExpression.
+     * @return a set of optimizer rules.
+     */
+    Set<ConnectorPlanOptimizer> getConnectorPlanOptimizers(TranslationContext<T> functionTranslator);
```

Comment:
Maybe we should just use `TranslationContext<T>` to avoid templated `ConnectorPlanOptimizerProvider<T>`. Check how other connector interfaces solve the problem. @hellium01 can help on this front.